### PR TITLE
Context component get

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #14800

Related to #14800.
This PR introduces a static method `Yii::get()`, which will pick a component from current module or application as a fallback.

### CODE REVIEW

**Review by samdark — APPROVED** [review]

**Review by sergeymakinen — APPROVED** [review]

**Comment by rob006** [comment]
Looks like BC break to me. If I have `request` component in my module, helpers will start using it instead of application request.

**Comment by klimov-paul** [comment]
> Looks like BC break to me. If I have request component in my module, helpers will start using it instead of application request.

Sorry, but this BC break is already started since #14350 merge. It is already too late for the complains.

**Comment by rob006** [comment]
@klimov-paul Can you give any example when #14350  breaks BC?

**Comment by klimov-paul** [comment]
> Can you give any example when #14350 breaks BC?

Any check or get of the component, which present in Application but absent in Module:

```php
$this->assertFalse($controller->module->has('errorHandler'));
```

**Comment by rob006** [comment]
I think that there is a big difference between "if module component doesn't exist, use fallback" and "pick some random component with requested name"...

**Comment by klimov-paul** [comment]
Since component tree traversal has been accepted there is no way back as it feature does not actually function otherwise.

**Comment by samdark** [comment]
@rob006 are there real cases when you define a component in the module w/ the same name as app component but do not use it?

**Comment by rob006** [comment]
@samdark I can create `user` component in my module, and I can use it, but it does not mean that this is `yii\web\User` and can be used in this context. This PR reserves some component names in every module and blindly assumes that everyone follows this new rule.

**Comment by klimov-paul** [comment]
@rob006, you are just repeating all arguments from #14350 here.
Its not me you need to convince - I was the one who stands against cretion of such twisted functionality at the first place.

This is your debates with @SamMousa. Once you both come to some sort of coclusion, let me know, will you?

**Comment by SamMousa** [comment]
I'll respond in #14800

File: framework/filters/AccessControl.php
Comment (SamMousa) [inline framework/filters/AccessControl.php]: Why not `$action->controller->module->get(...)` ?

File: framework/filters/AjaxFilter.php
Comment (SamMousa) [inline framework/filters/AjaxFilter.php]: What's the use case for a configurable `Request` object?
 (Understand if you have to keep it for backwards compatibility)

File: framework/filters/ContentNegotiator.php
Comment (SamMousa) [inline framework/filters/ContentNegotiator.php]: Why is this public? (understand if its BC)

File: framework/filters/Cors.php
Comment (SamMousa) [inline framework/filters/Cors.php]: Use `$action->controller->module->get('request')`

File: framework/filters/Cors.php
Comment (SamMousa) [inline framework/filters/Cors.php]: Use `$action->controller->module->get('response')`

File: framework/filters/Cors.php
Comment (klimov-paul) [reply framework/filters/Cors.php]: This is risk prone, since filter instance may be applied for the module (application) level not only for controller.

File: framework/filters/Cors.php
Comment (klimov-paul) [reply framework/filters/Cors.php]: This is risk prone, since filter instance may be applied for the module (application) level not only for controller.

**Comment by SamMousa** [comment]
In my opinion this PR should not (yet) be merged.
It uses a fundamentally bad approach of static service locator: `Yii::get()` which is exactly what we want to avoid.
Any behavior that is expected to be attached to a controller and thus receives and instance of `Action` should get stuff it needs from that action instead.

File: framework/filters/AccessControl.php
Comment (SamMousa) [reply framework/filters/AccessControl.php]: Hmm... still the parameter is an action right?
Alternatively we could check the owner of the behavior? (I've been considering implementing `ServiceLocator` on controllers *-))

**Comment by klimov-paul** [comment]
> Any behavior that is expected to be attached to a controller and thus receives and instance of Action should get stuff it needs from that action instead.

Controller behaviors (Filters) are likely to be dropped in favor of middleware in the future. Middleware can be aware of controller instance in general. Thus I suggest original issue should be closed.

**Comment by SamMousa** [comment]
So we don't change behaviors now since we expect them to be replaced anyway?
I can agree to that.

Alternatively I'd be open to introducing some thing like an internal API.
My main problem with introducing `Yii::get()` is that we can't remove it because it becomes part of the public API. 
If we mark it as internal (and we introduce that as a concept, not sure if we have) than it's fine for a temporary solution.

File: framework/filters/AccessControl.php
Comment (klimov-paul) [reply framework/filters/AccessControl.php]: Such check is already perfromed inside `Yii::get()` what is the sense of repeating it at every filter?

**Comment by klimov-paul** [comment]
> So we don't change behaviors now since we expect them to be replaced anyway? I can agree to that.

Since #15438. So you do not agree PSR middleware should appear at Yii?

> My main problem with introducing Yii::get() is that we can't remove it because it becomes part of the public API. If we mark it as internal (and we introduce that as a concept, not sure if we have) than it's fine for a temporary solution.

Just why is that: this new method is just a simple wrapper around existing and widely used `Yii::$app->get()` construction. Its only difference is usage tree traversal if controller instance is available - the feature you was so eager to create. This method simply reduces complex checks of controller instance existance inside custom code.

**Comment by SamMousa** [comment]
> Since #15438. So you do not agree PSR middleware should appear at Yii?

I think I misread; I am greatly in favor of PSR middleware and can agree that we don't change behaviors now per my issue and instead wait for PSR middleware and replace behaviors with that.

**Comment by SamMousa** [comment]
> Just why is that: this new method is just a simple wrapper around existing and widely used Yii::$app->get() construction. Its only difference is usage tree traversal if controller instance is available - the feature you was so eager to create. This method simply reduces complex checks of controller instance existance inside custom code.

Because it is another global variable / function. I'd prefer it if my application was the only singleton and anything that needs to be accessible is accessible via the app object, like `Yii::$app->..`.

Of course a stateless wrapper is not so bad, but it does promote bad usage since it should not be easy to use global variables, it should be easy to pass them into your scope.

File: framework/filters/AccessControl.php
Comment (SamMousa) [reply framework/filters/AccessControl.php]: The global static getter makes a lot of assumptions based on current structure of the framework; this means that that structure becomes more and more rigid since any change changes these global getters that are potentially used anywhere in a code base.

**Comment by samdark** [comment]
Either we're merging it (or modified version) now in 2.0.14 or not merging it into 2.0 at all.

**Comment by SamMousa** [comment]
I'm okay with not merging it based on @klimov-paul argument for moving stuff to PSR middleware in 2.1 (or 2.2).

**Comment by samdark** [comment]
Moved to 2.1. @klimov-paul feel free to close both issue and PR if you think they're not good for 2.1.

File: framework/filters/AccessControl.php
Comment (klimov-paul) [reply framework/filters/AccessControl.php]: This is exactly what you have pushed framework towards at #14350.
Instead of using explicit predefined methods for retrieval of the components, like [yii\web\Application::getRequest()](https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L158) or [yii\web\Application::getSession()](https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L176), you suggest usage of following constructions:

```
$controller->module->get('request'); // instead of `Yii::$app->getRequest()`

$controller->module->get('session'); // instead of `Yii::$app->getSession()`
```

as they allows triggering of tree traversal component search.

You have placed restrictions on component structure for the application and module yourself.
Why there are compains on it now?

**Comment by klimov-paul** [comment]
As far as I remember it was @rob006 who stands against the changeset. He should decide.
Still I do not understand why it should be moved to 2.1 as it is a simple polishing of #14350, which was accepted at 2.0.x.

**Comment by SamMousa** [comment]
In my opinion it is not polishing it is instead reversing.
Object instances should, whenever possible, not grab stuff from the global scope.
That's why i'm against it @klimov-paul; the goal of tree traversal is to have module local stuff, like controllers and actions (and behaviors attached to them) use module local components.

The goal is not to create yet another global magic retrieval system as you've implemented in this PR.

**Comment by klimov-paul** [comment]
Either this PR should be merged into 2.0.14 or this entire tree-traversal noncence introduced at #14350 should be reverted.

File: framework/filters/AccessControl.php
Comment (SamMousa) [reply framework/filters/AccessControl.php]: These are not dependent on the application having a single active controller or module; they are just dependent on direct relationships. These chains could (I'm not sure if they should) even be shortened by having a controller act like a service locator as well, but that's a whole different discussion.

I stand by my request and actually use tree traversal in my applications in production, it works and has valid use cases.

**Comment by klimov-paul** [comment]
> Object instances should, whenever possible, not grab stuff from the global scope.

That is impossible.
Having data and objects at the global scope is necessary. Without it there will be impossible to solve some common framework tasks.
For example: Database connection should be present as a single instance inside the application, as it should be configured (dsn, username, password) at the single place, and being opened only once (otherwise each connection open will consume redundant network resources).

Unless we pass application (or DI container) instance into each object created on the way, while resolving the request - there should be some way to get global data or objects, which are shared throughout the entire application.

**Comment by klimov-paul** [comment]
> The goal is not to create yet another global magic retrieval system as you've implemented in this PR.

You have already created with tree traverse. This PR simply wraps long check of controller existance for the tree traverse into a single method, so it can be reused easily.

**Comment by SamMousa** [comment]
> That is impossible.

I know that is why I said "whenever possible"; earlier in this thread I already mentioned AR as an example of where this approach doesn't work.

> You have already created with tree traverse. This PR simply wraps long check of controller existance for the tree traverse into a single method, so it can be reused easily.
The PR doesn't wrap a long check, it checks a different long path!
Going from action > controller > module is not the same as "global application object" > "current controller" > module... 
It might arrive at the same destination, but it is conceptually very different.

**Comment by klimov-paul** [comment]
The PR doesn't wrap a long check, it checks a different long path!
Going from action > controller > module is not the same as "global application object" > "current controller" > module...

The path will be the same as `Application::$controller` always stores current controller.
While not every piece of code may reference to the `Action` instance.

**Comment by klimov-paul** [comment]
`Yii::get()` just provides a quick access to the tree-traversal feature you have created., that is all.
It is a static call - yes, but it just replaces another static calls here - nothing more.
This PR is an exact manifistation of your dream of module, which can change the entire application flow using its own configuration,

**Comment by SamMousa** [comment]
> The path will be the same as Application::$controller always stores current controller.
While not every piece of code may reference to the Action instance.

That is my point exactly!
Suppose my main controller calls a function inside a module and that module uses your fancy `Yii::get()` then it will not get the write component!

**Comment by SamMousa** [comment]
I'm off for today; so if you have more arguments I'll respond tomorrow. If you want to make a choice without my approval then go ahead, in the end it's your framework not mine and I can't force you to do something.

**Comment by klimov-paul** [comment]
> Suppose my main controller calls a function inside a module and that module uses your fancy Yii::get() then it will not get the write component!

`Yii::get()` will pick up component from the current controller scope. If your module invocation does not switch controller you will have it correctly. If the module you refer to will use `Module::get()` inside it - it will pick its own component. Everything will work as it should.

**Comment by klimov-paul** [comment]
As I said I am just waiting for you and @rob006 to come to single conclusion as well as to the solid vision of the tree-traverse feature.
You have time to make the decision before 2.0.14. It is 1 week remaining.

**Comment by klimov-paul** [comment]
Although, no - I will not press this forward since there are concerns about changeset. Otherwise, it would be no better then decision made by #14350, where all this began.
Thus I will take the @samdark's [offer](https://github.com/yiisoft/yii2/pull/15342#issuecomment-363831905) closing both PR and issue.

**Pull request closed without merging**
